### PR TITLE
Implement Invisible (v3) ReCaptcha

### LINKED ISSUE #12: Feature: invisible recaptcha
Hi! Thanks a lot, this is very helpful library. I have question. I need implement invisible recaptcha from Google. Can i do this with your library? ty 😄

Not, its Google magic https://www.google.com/recaptcha/intro/comingsoon/invisible.html
Docs: https://developers.google.com/recaptcha/docs/invisible

The library can already be used. I tried to implement it, but I could not teach her to communicate with netteForms.js so I used your solution. My only chance is that you manage your experience.

### PR DESCRIPTION
**Pull request by JanGalek**
resolve https://github.com/contributte/reCAPTCHA/issues/12

I'm not sure with field must be required, 

Can be some test as bot ?

Maybe should will be create new type of field ?

### CODE REVIEW

**Comment by JanGalek** [comment]
I see https://product.hubspot.com/blog/quick-guide-invisible-recaptcha that I will must do more changes, pls no close, I will do rebase ;)

**Comment by mabar** [comment]
Could you separate it from basic recaptcha? I would like to have `addReCaptcha` and `addInvisibleReCaptcha`. Their use case is diferent and I can imagine that I need both of them.

ad. `data-callback`
- It's a diferent feature, please create separate PR for it.
- If I understand, it's a name of javascript function. It is view-specific and should not be in global configuration. You can set it to control same way as other atributes.

**Comment by JanGalek** [comment]
@mabar yes Its in plan with refactoring.

**Comment by JanGalek** [comment]
@mabar should be problem add js to docs ? (it will work with nittro too - iam testing with nittro).

File: src/Forms/InvisibleReCaptchaField.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	/**
+	 * @param string $message
+	 * @return static
+	 */
+	public function setMessage($message)
+	{
+		if ($this->configured === TRUE) {
+			throw new InvalidStateException('Please call setMessage() only once or don\'t pass $message over addReCaptcha()');
```
Comment (mabar) [inline src/Forms/InvisibleReCaptchaField.php]: addInvisibleReCaptcha

**Review by mabar — CHANGES_REQUESTED** [review]

**Comment by mabar** [comment]
I like that. Nice work.
I think javascript snippet is the only way here. But it could be done without jquery (example in [recaptcha docs](https://developers.google.com/recaptcha/docs/display#example) should be enough)

But I am curious if it could be even better. Did you think about an extension for netteForms.js?

**Comment by mabar** [comment]
Btw, please be sure that tests and qa pass. We have composer scripts for that https://github.com/contributte/reCAPTCHA/blob/master/composer.json#L36-L42
Coding style could be (partialy) fixed with `vendor/bin/codefixer`

If you are using Windows so I recommend you give a try to Linux subsystem (you can get Ubuntu from Windows store), because these scripts are not compatible with Windows.

**Comment by JanGalek** [comment]
@mabar I dont use netteForms, I'm using nittro, but can do that ;). It should be in docs or on gist or some another repository ?

**Comment by mabar** [comment]
@JanGalek I think you do and just don't know about it. netteForms.js is included in nittro :)
https://github.com/nittro/nittro/wiki/Forms#a-note-about-netteformsjs

Please create `assets/invisibleRecaptcha.js` and add note (and link to it) in docs.

**Comment by JanGalek** [comment]
@mabar changed, I'm not sure, if I added extension right.

**Comment by f3l1x** [comment]
Hi @JanGalek. Look's good to me. Have you test it? Could you please include a screenshot? You can include it into docs.

**Comment by JanGalek** [comment]
@f3l1x I tested on sandbox, now I'm going to test in my project with nittro. You can try it at https://test.gcore.cz/ (sandbox, not really validation username and password, you can write what you want).

**Comment by JanGalek** [comment]
@f3l1x ok I must little changed Nette form extension for compatibility with nittro. And screenshot was added.

reCAPTCHA

and small bonus, it automatic works with multiple forms which are on site without any manual changes ;)

**Comment by f3l1x** [comment]
Cool. What about the image in the right corner? It's required by rules by Google? It must be there?

File: src/Forms/InvisibleReCaptchaField.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+use Nette\Forms\Controls\HiddenField;
+use Nette\Forms\Form;
+use Nette\InvalidStateException;
+use Nette\Utils\Html;
+
+/**
+ * @author Milan Felix Sulc <sulcmil@gmail.com> | Jan Galek <jan.galek@troia-studio.cz>
+ */
+class InvisibleReCaptchaField extends HiddenField
+{
```
Comment (f3l1x) [inline src/Forms/InvisibleReCaptchaField.php:15]: What about create BaseReCaptcha field? It looks very similar to our other object.

**Comment by JanGalek** [comment]
@f3l1x yes this must be here.

>  As Google captures user information from your website, you'll need to warn your users by displaying their [Privacy policy](https://www.google.com/intl/en/policies/privacy/) and [Terms and conditions](https://www.google.com/intl/en/policies/privacy/).

File: src/Forms/InvisibleReCaptchaField.php (hunk 2 above)
Comment (JanGalek) [reply src/Forms/InvisibleReCaptchaField.php:15]: @f3l1x yes, but every field object extends from another object, should be it trait ?

Comment (f3l1x) [reply src/Forms/InvisibleReCaptchaField.php:15]: Oh, I didn't notice. Just keep it. :-) It's ok.

**Comment by JanGalek** [comment]
@f3l1x @mabar ping

**Comment by f3l1x** [comment]
Cool. Thank you for the great job.

**Comment by f3l1x** [comment]
Do you wanna join us at @contributte? I think it would be great to have you.

**Comment by JanGalek** [comment]
@f3l1x yes, it would be a pleasure, I will write you to slack (pehapkari)

**Comment by natocTo** [comment]
Good job. I'm Looking forward to 3.1 release.

**Pull request merged**